src/Microsoft.AspNetCore.Sockets.Common/Internal/Formatters/TextMessageParser.cs

```diff
@@ -0,0 +1,240 @@
+            if (lengthSpan == null)
+            {
+                // Insufficient data
+                return false;
+            }
+
+            // Parse the length
+            if (!PrimitiveParser.TryParseInt32(lengthSpan.Value, out var length, out var consumedByLength, encoder: TextEncoder.Utf8) || consumedByLength < lengthSpan.Value.Length)
+            {
+                if (TextEncoder.Utf8.TryDecode(lengthSpan.Value, out var lengthString, out _))
+                {
+                    throw new FormatException($"Invalid length: '{lengthString}'");
+                }
+
+                throw new FormatException("Invalid length!");
+            }
+
+            _state.Length = length;
+            _state.Phase = ParsePhase.LengthComplete;
+            return true;
+        }
+
+        private bool TryReadDelimiter(ref BytesReader buffer, char delimiter, ParsePhase nextPhase, string field)
+        {
+            if (buffer.Unread.Length == 0)
+            {
+                return false;
+            }
+
+            if (buffer.Unread[0] != delimiter)
+            {
+                throw new FormatException($"Missing delimiter '{delimiter}' after {field}");
+            }
+            buffer.Advance(1);
+
+            _state.Phase = nextPhase;
+            return true;
+        }
+
+        private bool TryReadType(ref BytesReader buffer)
+        {
+            if (buffer.Unread.Length == 0)
+            {
+                return false;
+            }
+
+            if (!TryParseType(buffer.Unread[0], out _state.MessageType))
+            {
+                throw new FormatException($"Unknown message type: '{(char)buffer.Unread[0]}'");
+            }
+
+            buffer.Advance(1);
+            _state.Phase = ParsePhase.TypeComplete;
+            return true;
+        }
+
+        private void ReadPayload(ref BytesReader buffer)
+        {
+            if (_state.Payload == null)
+            {
+                _state.Payload = new byte[_state.Length];
+            }
+
+            if (_state.Read >= _state.Length)
+            {
+                _state.Payload = DecodePayload(_state.Payload);
+
+                _state.Phase = ParsePhase.PayloadComplete;
+            }
+            else
+            {
+                // Copy as much as possible from the Unread buffer
+                var toCopy = Math.Min(_state.Length, buffer.Unread.Length);
+                buffer.Unread.Slice(0, toCopy).CopyTo(_state.Payload.Slice(_state.Read));
+                _state.Read += toCopy;
+                buffer.Advance(toCopy);
+            }
+        }
+
+        private byte[] DecodePayload(byte[] inputPayload)
+        {
+            if (_state.MessageType == MessageType.Binary && inputPayload.Length > 0)
+            {
+                // Determine the output size
+                // Every 4 Base64 characters represents 3 bytes
+                var decodedLength = (inputPayload.Length / 4) * 3;
```

Comment:
It makes some sense that it wouldn't be there yet because you can't do it with the input length alone, you need to look at the padding bytes to determine the actual length.